instrumentor/controllers/sourceinstrumentation/common.go

```diff
@@ -0,0 +1,184 @@
+package sourceinstrumentation
+
+import (
+	"context"
+	"errors"
+
+	v1 "k8s.io/api/apps/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/odigos-io/odigos/api/k8sconsts"
+	"github.com/odigos-io/odigos/api/odigos/v1alpha1"
+	sourceutils "github.com/odigos-io/odigos/k8sutils/pkg/source"
+	k8sutils "github.com/odigos-io/odigos/k8sutils/pkg/utils"
+	"github.com/odigos-io/odigos/k8sutils/pkg/workload"
+)
+
+func syncNamespaceWorkloads(
+	ctx context.Context,
+	k8sClient client.Client,
+	runtimeScheme *runtime.Scheme,
+	namespace string) (ctrl.Result, error) {
+
+	collectiveRes := ctrl.Result{}
+	var errs error
+	for _, kind := range []k8sconsts.WorkloadKind{
+		k8sconsts.WorkloadKindDaemonSet,
+		k8sconsts.WorkloadKindDeployment,
+		k8sconsts.WorkloadKindStatefulSet,
+	} {
+		workloadObjects := workload.ClientListObjectFromWorkloadKind(kind)
+		err := k8sClient.List(ctx, workloadObjects, client.InNamespace(namespace))
+		if err != nil {
+			errs = errors.Join(errs, err)
+			continue
+		}
+
+		objects := make([]client.Object, 0)
+		switch obj := workloadObjects.(type) {
+		case *v1.DeploymentList:
+			for _, dep := range obj.Items {
+				objects = append(objects, &dep)
+			}
+		case *v1.DaemonSetList:
+			for _, ds := range obj.Items {
+				objects = append(objects, &ds)
+			}
+		case *v1.StatefulSetList:
+			for _, ss := range obj.Items {
+				objects = append(objects, &ss)
+			}
+		}
+
+		for _, obj := range objects {
+			res, err := syncWorkload(ctx, k8sClient, runtimeScheme, obj)
```

Comment:
When iterating a list, the objects might change from the time the api server returned until the object is being consumed.
I suggest to `Get` a fresh copy of the object before using it.